cpp/src/arrow/dataset/expression.cc

```diff
@@ -681,7 +681,7 @@ Result<Expression> FoldConstants(Expression expr) {
         if (std::all_of(call->arguments.begin(), call->arguments.end(),
                         [](const Expression& argument) { return argument.literal(); })) {
           // all arguments are literal; we can evaluate this subexpression *now*
-          static const Datum ignored_input;
+          static const Datum ignored_input = Datum{};
```

Comment:
Perhaps we can give Datum a default constructor?

Reply:
https://github.com/bkietz/arrow/blob/17f2df78f1fa623d05dd01f245f8f65f828f510d/cpp/src/arrow/datum.h#L118-L119

Datum has one already; resolution of that constructor is just failing here

Reply:
Uh.

Reply:
precisely